### PR TITLE
alerts: Extend MimirAllocatingTooMuchMemory for read-write deployment

### PR DESCRIPTION
**Pull request by jhesketh**
#### What this PR does

Extend MimirAllocatingTooMuchMemory for read-write deployment

Also check mimir-write which contains the ingester, and mimir-backend which is critical (and contains AlertManager).

#### Which issue(s) this PR fixes or relates to

Partial fixes https://github.com/grafana/mimir/issues/3366

#### Checklist

- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -18,7 +18,8 @@
 ### Mixin
 
 * [ENHANCEMENT] Alerts: Added `MimirIngesterInstanceHasNoTenants` alert that fires when an ingester replica is not receiving write requests for any tenant. #3681
-* [BUGFIX] Alerts: Fixed `MemoryMapAreasTooHigh` alert when Mimir is deployed in read-write mode. #3626
+* [ENHANCEMENT] Alerts: Extended `MimirAllocatingTooMuchMemory` to check read-write deployment containers. #3710
+* [BUGFIX] Alerts: Fixed `MimirMemoryMapAreasTooHigh` alert when Mimir is deployed in read-write mode. #3626
```
Comment (pracucci) [inline CHANGELOG.md:22]: Thanks for fixing this entry too!

File: operations/mimir-mixin/alerts/alerts.libsonnet [hunk 2]
```diff
@@ -557,7 +557,7 @@ local utils = import 'mixin-utils/utils.libsonnet';
           },
           annotations: {
             message: |||
-              Ingester %(alert_instance_variable)s in %(alert_aggregation_variables)s is using too much memory.
+              Container %(alert_instance_variable)s in %(alert_aggregation_variables)s is using too much memory.
```
Comment (pracucci) [inline operations/mimir-mixin/alerts/alerts.libsonnet:560]: I would generalise it. "Instance" is also the naming we use in the doc.

```suggestion
              Ingester %(alert_instance_variable)s in %(alert_aggregation_variables)s is using too much memory.
```

File: operations/mimir-mixin/alerts/alerts.libsonnet [hunk 3]
```diff
@@ -578,7 +578,7 @@ local utils = import 'mixin-utils/utils.libsonnet';
...
```
Comment (pracucci) [inline operations/mimir-mixin/alerts/alerts.libsonnet:581]: Same here.

```suggestion
              Instance %(alert_instance_variable)s in %(alert_aggregation_variables)s is using too much memory.
```

File: operations/mimir-mixin/alerts/alerts.libsonnet [hunk 4]
```diff
@@ -567,9 +567,9 @@ local utils = import 'mixin-utils/utils.libsonnet';
             (
               # We use RSS instead of working set memory because of the ingester's extensive usage of mmap.
               # See: https://github.com/grafana/mimir/issues/2466
-              container_memory_rss{container="ingester"}
+              container_memory_rss{container=~"(ingester|mimir-write|mimir-backend)"}
```
Comment (pracucci) [inline operations/mimir-mixin/alerts/alerts.libsonnet:570]: Don't hardcode container names, but build the regex from `$._config.container_names` please.

_Same comment applies in other places._

**Review by pracucci — COMMENTED** [review]
LGTM (modulo a couple of nits), thanks!

File: operations/mimir-mixin/alerts/alerts.libsonnet (hunk 2 above)
Comment (jhesketh) [reply operations/mimir-mixin/alerts/alerts.libsonnet:560]: You've accidentally put Ingester as a suggestion so I'll push a new commit.

Comment (pracucci) [reply operations/mimir-mixin/alerts/alerts.libsonnet]: Doh! Finger memory is a bad thing :(

File: operations/mimir-mixin/alerts/alerts.libsonnet [hunk 5]
```diff
@@ -550,18 +550,22 @@ local utils = import 'mixin-utils/utils.libsonnet';
...
               # See: https://github.com/grafana/mimir/issues/2466
-              container_memory_rss{container="ingester"}
+              container_memory_rss{container=~"(%(ingester)s|%(mimir_write)s|%(mimir_backend)s)"}
                 /
-              ( container_spec_memory_limit_bytes{container="ingester"} > 0 )
+              ( container_spec_memory_limit_bytes{container=~"(%(ingester)s|%(mimir_write)s|%(mimir_backend)s)"} > 0 )
             ) > 0.65
-          |||,
+          ||| % {
+            ingester: $._config.instance_names.ingester,
```
Comment (pracucci) [inline operations/mimir-mixin/alerts/alerts.libsonnet:558]: I think it should be `container_names` not `instance_names`.

_Comment applies everywhere else._

Comment (jhesketh) [reply operations/mimir-mixin/alerts/alerts.libsonnet:558]: So I think in the case of the `ingester`, we would be replicating the logic of `helmCompatibleMatcher`, `baremetalCompatibleMatcher`, and `instanceMatcher`, to get the same regex that we can already get in `instance_names`.

As for mimir_backend, in `container_names` it uses `componentGroups.backend` meaning that it'd create a regex for `['query_scheduler', 'ruler_query_scheduler', 'ruler', 'store_gateway', 'compactor', 'alertmanager', 'overrides_exporter', 'mimir_backend']` which would over-include services for microservice deployment mode.

Comment (pracucci) [reply operations/mimir-mixin/alerts/alerts.libsonnet:558]: > So I think in the case of the ingester, we would be replicating the logic of helmCompatibleMatcher, baremetalCompatibleMatcher, and instanceMatcher, to get the same regex that we can already get in instance_names.

I'm not sure about it. Few things:
1. Helm: the container name shouldn't change, so should always be `ingester` (the `helmCompatibleMatcher()` is used to match the deployment/pod which can have the `mimir-` prefix)
2. Baremetal: there are no `container_*` metrics anyway there

Comment (jhesketh) [reply operations/mimir-mixin/alerts/alerts.libsonnet]: okay, makes sense. I had to add an extra selector for just the read-write containers but it looks neater now.

**Review by pracucci — APPROVED** [review]
LGTM, thanks!

**Pull request merged**